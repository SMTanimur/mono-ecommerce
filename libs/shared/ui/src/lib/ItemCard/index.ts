import { ItemCard, ItemCardProps } from './ItemCard';

export { ItemCard };

export type { ItemCardProps };
