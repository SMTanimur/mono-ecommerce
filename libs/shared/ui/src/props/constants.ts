import { ButtonProps } from '@ecommerce/shared/ui/Button';
import {CartItemProps} from "@ecommerce/shared/ui/CartItem";
import {ProductSidebarProps} from "@ecommerce/shared/ui/ProductSidebar";
import {CartSidebarProps} from "@ecommerce/shared/ui/CartSidebar";
import {CategoryCardProps} from "@ecommerce/shared/ui/CategoryCard";
import {IconProps} from "@ecommerce/shared/ui/Icon";
import {ItemCardProps} from "@ecommerce/shared/ui/ItemCard";
import {NavButtonProps} from "@ecommerce/shared/ui/NavButton";
import {ProductGalleryProps} from "@ecommerce/shared/ui/ProductGallery";

import cat from './cat.jpg';
import cat2 from './cat2.jpeg';

console.log(cat2)

export const PRODUCT: ProductSidebarProps['product'] = {
  slug: 'some-name',
  price: 65,
  name: 'Some name',
  images: {
    data: [
      {
        attributes: {
          url: cat,
        },
      },
      {
        attributes: {
          url: cat2,
        },
      },
    ],
  },
  smallDescription: 'Small Description',
  description: 'Description',
  information: [
    {
      title: 'Title1',
      content: 'content1',
    },
  ],
};

export const ITEM: ItemCardProps['item'] = {
  slug: 'cat-cat',
  price: 65,
  name: 'cat cat',
  images: {
    data: [
      {
        attributes: {
          url: cat,
        },
      },
      {
        attributes: {
          url: cat2,
        },
      },
    ],
  },
  smallDescription: 'smallDescription',
  description: 'description',
};

export const BUTTON_DEFAULT_PROPS: ButtonProps = {
  children: 'hello',
  tag: 'button',
  size: 'base',
  type: 'primary',
  round: 'full',
};

export const CART_ITEM_DEFAULT_PROPS: CartItemProps = {
  inCartAmount: 1,
  item: PRODUCT,
  handleCartMinusClick: () => {
    console.log('cart minus');
  },
  handleCartPlusClick: () => {
    console.log('cart plus');
  },
  handleRemoveFromCart: () => {
    console.log('cart remove');
  },
  toggleFavorite: () => {
    console.log('favorite');
  },
};

export const CART_SIDEBAR_DEFAULT_PROPS: CartSidebarProps = {
  cartTotal: 100,
  isUser: true,
  handleCheckoutClick: async () => {
    console.log('checkout');
  },
  isLoading: false,
};

export const CATEGORY_CARD_DEFAULT_PROPS: CategoryCardProps = {
  category: {
    smallDescription:
      'smalll des',
    image: {
      data: {
        attributes: {
          url: cat,
        },
      },
    },
    slug: '1',
  },
};
export const ICON_DEFAULT_PROPS: IconProps = {
  icon: 'cart',
};

export const ITEM_CARD_DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleCartClick: () => console.log('cart clicked'),
  handleAmountCartClick: () => console.log('amount cart clicked'),
  handleFavoriteClick: () => console.log('favorite clicked'),
};
export const NAV_BUTTON_DEFAULT_PROPS: NavButtonProps = {
  type: 'heart',
};
export const PRODUCT_GALLERY_DEFAULT_PROPS: ProductGalleryProps = {
  images: [
    {
      attributes: {
        url: cat,
      },
    },
    {
      attributes: {
        url: cat2,
      },
    },
  ],
};
