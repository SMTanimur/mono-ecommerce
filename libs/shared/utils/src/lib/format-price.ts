export const formatPrice = (price: number) =>
  `${Math.max(price, 0).toFixed(2)}$`;
