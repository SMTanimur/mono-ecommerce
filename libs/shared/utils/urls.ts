export const getApiUrl = (path = ''): string =>
  `${process.env['NX_STRAPI_IMG_URL'] ?? ''}${path}`;

export const getImageUrl = (path = ''): string =>
  `${process.env['NX_STRAPI_URL'] ?? 'http://localhost:4200'}${path}`;

export const createItemHref = (id: string): string => 
    `/i/${id}`;
