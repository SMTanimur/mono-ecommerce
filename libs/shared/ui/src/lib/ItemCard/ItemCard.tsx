import React, { useMemo, useState } from 'react';

import clsx from 'clsx';
import Link from 'next/link';

import type { Product } from '@ecommerce/shared/graphql/refactored-types';
import { handleBlur } from '@ecommerce/shared/utils/handle-blur';
import { createItemHref } from '@ecommerce/shared/utils/url';

import { Amount } from './LocalComponents/Amount';
import { Cart } from './LocalComponents/Cart';
import { CoverImage } from './LocalComponents/CoverImage';
import { Favorite } from './LocalComponents/Favorite';
import { Information } from './LocalComponents/Information';

export interface ItemCardProps {
    item: Product;
    handleCartClick: () => void;
    handleAmountCartClick: () => void;
    handleFavoriteClick: () => void;
    size?: typeof SIZES[number];
    elementProps?: JSX.IntrinsicElements['div'];
    isFavorite?: boolean;
    inCartAmount?: number;
}

const SIZES = ['base', 'lg', 'parent'] as const;

export function createLabel(item: Product) {
    const { name, smallDescription, price } = item;

    return `${name}, ${smallDescription}, price: ${price}`;
}

export const ItemCard: React.FC<ItemCardProps> = (props) => {
    const {
        item,
        elementProps,
        size = 'parent',
        handleCartClick,
        handleAmountCartClick,
        handleFavoriteClick,
        isFavorite = false,
        inCartAmount = 0,
    } = props;

    const [{ hover, focus }, setFocusHover] = useState({
        focus: false,
        hover: false,
    });

    const hovered = useMemo(() => focus || hover, [focus, hover]);
    const href = useMemo(() => createItemHref(item.slug), [item.slug]);

    return (
        <div
            className={clsx(
                'relative',
                'flex flex-col',
                'list-none',
                size === 'base' && 'w-60',
                size === 'lg' && 'w-72',
                size === 'parent' && 'w-full h-full',
            )}
            {...elementProps}
            onMouseOver={() =>
                setFocusHover((prev) => ({ ...prev, hover: true }))
            }
            onMouseLeave={() =>
                setFocusHover((prev) => ({ ...prev, hover: false }))
            }
            onFocus={() => setFocusHover((prev) => ({ ...prev, focus: true }))}
            onBlur={(e) => {
                handleBlur(e)(
                    () =>
                        void setFocusHover((prev) => ({
                            ...prev,
                            focus: false,
                        })),
                );
            }}
            data-testid="item-card"
        >
            <Link href={href}>
                <a className="mb-1" title={createLabel(item)}>
                    <div
                        className={clsx(
                            'relative w-full mb-2',
                            'object-cover bg-gray-100',
                            size === 'parent' && 'aspect-square',
                            size === 'base' && 'h-72',
                            size === 'lg' && 'h-80',
                        )}
                    >
                        <CoverImage product={item} hovered={hovered} />
                    </div>
                    <Information product={item} hovered={hovered} />
                </a>
            </Link>
            <div className="flex gap-2 md:(absolute right-0 bottom-0) ">
                <div className={clsx('relative', !hovered && 'sm:(opacity-0)')}>
                    <Cart handleClick={handleCartClick} />
                    {inCartAmount > 0 && (
                        <Amount
                            handleClick={handleAmountCartClick}
                            inCartAmount={inCartAmount}
                        />
                    )}
                </div>
                <Favorite
                    handleClick={handleFavoriteClick}
                    hovered={hovered}
                    isFavorite={isFavorite}
                />
            </div>
        </div>
    );
};
