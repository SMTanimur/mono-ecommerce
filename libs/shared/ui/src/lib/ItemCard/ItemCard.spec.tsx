import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';

import { ITEM_CARD_DEFAULT_PROPS } from '@ecommerce/shared/props/constants';
import { formatPrice } from '@ecommerce/shared/utils/format-price';

import { ItemCard, createLabel } from './ItemCard';

describe('ItemCard', () => {
  let element: HTMLElement;
  const props = ITEM_CARD_DEFAULT_PROPS;

  beforeEach(() => {
    const { baseElement } = render(<ItemCard {...props} />);

    element = baseElement;
  });

  it('should render successfully', () => {
    const tree = create(<ItemCard {...props} />).toJSON();
    expect(tree).toMatchSnapshot();

    expect(element).toBeTruthy();
  });

  it('main elements are exists and properly formatted', () => {
    const { price, name, smallDescription } = props.item;

    const properPrice = formatPrice(price);

    screen.getByText(name);
    screen.getByText(properPrice);
    screen.getByText(smallDescription);
  });

  it('proper link title', () => {
    const { price, name, smallDescription } = props.item;

    const title = createLabel(props.item);
    expect(title).toBe(`${name}, ${smallDescription}, price: ${price}`);

    screen.queryByTitle(title);
  });

  it('proper state on focus/hover', () => {
    const { name } = props.item;

    const container = screen.getByTestId('item-card');

    const figcaption = screen.getByText(name);
    expect(figcaption).not.toHaveClass('underline');

    const buttonCartContainer = screen.getByRole('button', {
      name: 'add to cart',
    }).parentElement;
    expect(buttonCartContainer).toHaveClass('sm:(opacity-0)');

    const buttonFavorite = screen.getByRole('button', {
      name: 'add to favorite',
    });
    expect(buttonFavorite).toHaveClass('sm:(opacity-0)');

    const secondImage = screen.getByTestId('second-image');
    expect(secondImage).toHaveClass('opacity-0');

    fireEvent.mouseOver(container);

    expect(figcaption).toHaveClass('underline');
    expect(buttonCartContainer).not.toHaveClass('sm:(opacity-0)');
    expect(buttonFavorite).not.toHaveClass('sm:(opacity-0)');
    expect(secondImage).toHaveClass('opacity-100');
  });
});
