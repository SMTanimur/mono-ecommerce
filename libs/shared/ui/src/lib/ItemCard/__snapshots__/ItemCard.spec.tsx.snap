// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCard should render successfully 1`] = `
<div
  className="relative flex flex-col list-none w-60"
  data-testid="item-card"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <a
    className="mb-1"
    href="/i/cat-cat"
    onClick={[Function]}
    onMouseEnter={[Function]}
    title="cat cat, smallDescription, price: 65"
  >
    <div
      className="relative w-full mb-2 object-cover bg-gray-100 h-72"
    >
      Next image stub
      <div
        aria-hidden="true"
        className="absolute left-0 top-0 h-full w-full transition-opacity opacity-0"
        data-testid="second-image"
      >
        Next image stub
      </div>
    </div>
    <div
      className="flex flex-col gap-1"
    >
      <span
        className="text-sm font-semibold uppercase sm:(text-base)"
      >
        cat cat
      </span>
      <span
        className="text-xs sm:(text-sm max-w-3/5)"
      >
        smallDescription
      </span>
      <span
        className="font-bold text-md sm:(text-xl)"
      >
        65.00$
      </span>
    </div>
  </a>
  <div
    className="flex gap-2 md:(absolute right-0 bottom-0) "
  >
    <div
      className="relative sm:(opacity-0)"
    >
      <button
        className="inline-block cursor-pointer transform-gpu transition-all active:scale-92 rounded-full p-3 text-white bg-sky-600 hover:(bg-sky-700) active:(bg-sky-800) flex items-center"
        onClick={[Function]}
        title="add to cart"
        type="button"
      >
        <div
          aria-hidden="true"
          className="relative flex items-center select-none pointer-events-none h-4 w-4  sm:(h-6 w-6)"
        >
          <svg
            className="feather feather-shopping-cart "
            fill="none"
            height={24}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width={24}
          >
            <g
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<circle cx=\\"9\\" cy=\\"21\\" r=\\"1\\"></circle><circle cx=\\"20\\" cy=\\"21\\" r=\\"1\\"></circle><path d=\\"M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6\\"></path>",
                }
              }
            />
          </svg>
        </div>
      </button>
    </div>
    <button
      className="inline-block cursor-pointer transform-gpu transition-all active:scale-92 rounded-full p-3 hover:(bg-gray-100) active:(bg-gray-300) flex items-center sm:(opacity-0)"
      onClick={[Function]}
      title="add to favorite"
      type="button"
    >
      <div
        aria-hidden="true"
        className="relative flex items-center select-none pointer-events-none h-4 w-4  sm:(h-6 w-6)"
      >
        <svg
          className="feather feather-heart "
          fill="none"
          height={24}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={24}
        >
          <g
            dangerouslySetInnerHTML={
              Object {
                "__html": "<path d=\\"M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z\\"></path>",
              }
            }
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;
